### PR TITLE
feat(functions): Implement "IS [NOT] DISTINCT FROM" operator"

### LINKED ISSUE #6046: Implement `IS [NOT] DISTINCT FROM` operator
`IS [NOT] DISTINCT FROM` operator is used to compare two expression with awareness of nullability.

For example, the result of `1 IS DISTINCT FROM NULL` is `true` and the result of `1 IS NOT DISTINCT FROM NULL` is `false`, in contrast the result of both `1 <> NULL` and `1 = NULL` is a special falsy value because `NULL` is a unknown value in SQL standard.

The table of value with `IS DISTINCT FROM`:

|lhs IS DISTINCT FROM rhs|1|2|NULL|
|---|--|--|--|
|1|FALSE|TRUE|TRUE|
|2|TRUE|FALSE|TRUE|
|NULL|TRUE|TRUE|FALSE|

The table of value with `IS NOT DISTINCT FROM`:

|lhs IS NOT DISTINCT FROM rhs|1|2|NULL|
|---|--|--|--|
|1|TRUE|FALSE|FALSE|
|2|FALSE|TRUE|FALSE|
|NULL|FALSE|FALSE|TRUE|

### PR DESCRIPTION
**Pull request by JialuGong**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support `IS [NOT] distinct function` in planner_v2.

## Changelog

- New Feature

## Related Issues

Fixes #6046

### CODE REVIEW

File: common/ast/src/parser/expr.rs [hunk 1]
```diff
@@ -291,6 +294,7 @@ impl<'a, I: Iterator<Item = WithSpan<'a, ExprElement<'a>>>> PrattParser<I> for E
             ExprElement::MapAccess { .. } => Affix::Postfix(Precedence(10)),
             ExprElement::IsNull { .. } => Affix::Postfix(Precedence(17)),
             ExprElement::Between { .. } => Affix::Postfix(Precedence(BETWEEN_PREC)),
+            ExprElement::IsDistinctFrom { .. } => Affix::Infix(Precedence(20), Associativity::Left),
```
Comment (andylokandy) [inline common/ast/src/parser/expr.rs:297]: Is 20 BETWEEN_PREC? (IIRC)

File: common/ast/src/parser/expr.rs [hunk 2]
```diff
@@ -873,6 +888,7 @@ pub fn expr_element(i: Input) -> IResult<WithSpan<ExprElement>> {
             | #trim_from : "`TRIM([(BOTH | LEADEING | TRAILING) ... FROM ...)`"
             | #nullif: "`NULLIF(..., ...)`"
             | #ifnull: "`IFNULL(..., ...)`"
+            | #is_distinct_from :"`... IS [NOT] DISTINCT FROM ...`"
```
Comment (andylokandy) [inline common/ast/src/parser/expr.rs:891]:
```suggestion
            | #is_distinct_from : "`... IS [NOT] DISTINCT FROM ...`"
```

**Review by andylokandy — APPROVED** [review]

File: common/ast/src/parser/expr.rs (hunk 1 above)
Comment (JialuGong) [reply common/ast/src/parser/expr.rs:297]: Yes
https://github.com/datafuselabs/databend/blob/c0d05f18ee4dea6c662539d8e0414ad87e938a6f/common/ast/src/parser/expr.rs#L36
 I will fix it ~

File: query/src/sql/planner/semantic/type_check.rs [hunk 3]
```diff
@@ -161,6 +161,63 @@ impl<'a> TypeChecker<'a> {
...
+                    span,
+                    expr: right.clone(),
+                    not: false,
+                });
+                let op = if *not {
+                    BinaryOperator::Eq
+                } else {
+                    BinaryOperator::NotEq
+                };
+                self.resolve_function(
```
[the remark this answers is not in the document]
Comment (JialuGong) [reply query/src/sql/planner/semantic/type_check.rs:185]: Okay, I will fix it ~

File: common/ast/src/parser/expr.rs [hunk 4]
```diff
@@ -873,6 +889,7 @@ pub fn expr_element(i: Input) -> IResult<WithSpan<ExprElement>> {
             | #trim_from : "`TRIM([(BOTH | LEADEING | TRAILING) ... FROM ...)`"
             | #nullif: "`NULLIF(..., ...)`"
             | #ifnull: "`IFNULL(..., ...)`"
+            | #is_distinct_from : "`... IS [NOT] DISTINCT FROM ...`"
```
Comment (Xuanwo) [inline common/ast/src/parser/expr.rs:892]:
```suggestion
            | #is_distinct_from: "`... IS [NOT] DISTINCT FROM ...`"
```

**Comment by Xuanwo** [comment]
This PR could affected by https://github.com/datafuselabs/databend/pull/6176, FYI

File: query/src/sql/planner/semantic/type_check.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JialuGong) [reply query/src/sql/planner/semantic/type_check.rs:185]: It seems that `resolve_function` accepts arguments which are `Expr` enum,  how every the `f` is `(Scalar, DataTypeImpl)`, may we could use `FunctionCall`  or add `AssumeNotNullableExpr` in `Scalar` enum which can accept `Scalar` args.

**Comment by BohuTANG** [comment]
```
[ FAIL ] - result differs with:
--- /Users/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0059_function_is_distinct_from.result	2022-06-24 02:01:26.000000000 +0000
+++ /Users/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0059_function_is_distinct_from.stdout	2022-06-24 02:06:37.000000000 +0000
@@ -1,40 +1,30 @@
 ==is_distinct_from_non_nullable_const_values==
-1
-0
-1
-0
+ERROR 1105 (HY000) at line 4: Code: 1005, displayText = Unsupported expression: 1 IS DISTINCT FROM 2, type: IsDistinctFrom(Value(Number("1", false)), Value(Number("2", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 5: Code: 1005, displayText = Unsupported expression: 1 IS DISTINCT FROM 1, type: IsDistinctFrom(Value(Number("1", false)), Value(Number("1", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 6: Code: 1005, displayText = Unsupported expression: 'a' IS DISTINCT FROM 'b', type: IsDistinctFrom(Value(SingleQuotedString("a")), Value(SingleQuotedString("b"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 7: Code: 1005, displayText = Unsupported expression: 'a' IS DISTINCT FROM 'a', type: IsDistinctFrom(Value(SingleQuotedString("a")), Value(SingleQuotedString("a"))) (while in analyze select projection).
 ==is_not_distinct_from_non_nullable_const_values==
```

File: query/src/sql/planner/semantic/type_check.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sundy-li) [reply query/src/sql/planner/semantic/type_check.rs:187]: I think to use `Expr:: FunctionCall ` is better.

**Comment by BohuTANG** [comment]
```
02_0059_function_is_distinct_from:                                      [ FAIL ] - result differs with:
--- /workspace/tests/suites/0_stateless/02_function/02_0059_function_is_distinct_from.result	2022-06-24 08:48:59.522499991 +0000
+++ /workspace/tests/suites/0_stateless/02_function/02_0059_function_is_distinct_from.stdout	2022-06-24 08:53:40.112755978 +0000
@@ -1,40 +1,30 @@
+ERROR 1105 (HY000) at line 4: Code: 1005, displayText = Unsupported expression: 1 IS DISTINCT FROM 2, type: IsDistinctFrom(Value(Number("1", false)), Value(Number("2", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 5: Code: 1005, displayText = Unsupported expression: 1 IS DISTINCT FROM 1, type: IsDistinctFrom(Value(Number("1", false)), Value(Number("1", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 6: Code: 1005, displayText = Unsupported expression: 'a' IS DISTINCT FROM 'b', type: IsDistinctFrom(Value(SingleQuotedString("a")), Value(SingleQuotedString("b"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 7: Code: 1005, displayText = Unsupported expression: 'a' IS DISTINCT FROM 'a', type: IsDistinctFrom(Value(SingleQuotedString("a")), Value(SingleQuotedString("a"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 11: Code: 1005, displayText = Unsupported expression: 1 IS NOT DISTINCT FROM 2, type: IsNotDistinctFrom(Value(Number("1", false)), Value(Number("2", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 12: Code: 1005, displayText = Unsupported expression: 1 IS NOT DISTINCT FROM 1, type: IsNotDistinctFrom(Value(Number("1", false)), Value(Number("1", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 13: Code: 1005, displayText = Unsupported expression: 'a' IS NOT DISTINCT FROM 'b', type: IsNotDistinctFrom(Value(SingleQuotedString("a")), Value(SingleQuotedString("b"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 14: Code: 1005, displayText = Unsupported expression: 'a' IS NOT DISTINCT FROM 'a', type: IsNotDistinctFrom(Value(SingleQuotedString("a")), Value(SingleQuotedString("a"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 17: Code: 1005, displayText = Unsupported expression: 1 IS DISTINCT FROM NULL, type: IsDistinctFrom(Value(Number("1", false)), Value(Null)) (while in analyze select projection).
+ERROR 1105 (HY000) at line 18: Code: 1005, displayText = Unsupported expression: NULL IS DISTINCT FROM 1, type: IsDistinctFrom(Value(Null), Value(Number("1", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 19: Code: 1005, displayText = Unsupported expression: 'a' IS DISTINCT FROM NULL, type: IsDistinctFrom(Value(SingleQuotedString("a")), Value(Null)) (while in analyze select projection).
+ERROR 1105 (HY000) at line 20: Code: 1005, displayText = Unsupported expression: NULL IS DISTINCT FROM 'b', type: IsDistinctFrom(Value(Null), Value(SingleQuotedString("b"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 21: Code: 1005, displayText = Unsupported expression: NULL IS DISTINCT FROM NULL, type: IsDistinctFrom(Value(Null), Value(Null)) (while in analyze select projection).
+ERROR 1105 (HY000) at line 24: Code: 1005, displayText = Unsupported expression: 1 IS NOT DISTINCT FROM NULL, type: IsNotDistinctFrom(Value(Number("1", false)), Value(Null)) (while in analyze select projection).
+ERROR 1105 (HY000) at line 25: Code: 1005, displayText = Unsupported expression: NULL IS NOT DISTINCT FROM 1, type: IsNotDistinctFrom(Value(Null), Value(Number("1", false))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 26: Code: 1005, displayText = Unsupported expression: 'a' IS NOT DISTINCT FROM NULL, type: IsNotDistinctFrom(Value(SingleQuotedString("a")), Value(Null)) (while in analyze select projection).
+ERROR 1105 (HY000) at line 27: Code: 1005, displayText = Unsupported expression: NULL IS NOT DISTINCT FROM 'b', type: IsNotDistinctFrom(Value(Null), Value(SingleQuotedString("b"))) (while in analyze select projection).
+ERROR 1105 (HY000) at line 28: Code: 1005, displayText = Unsupported expression: NULL IS NOT DISTINCT FROM NULL, type: IsNotDistinctFrom(Value(Null), Value(Null)) (while in analyze select projection).
+ERROR 1105 (HY000) at line 33: Code: 1005, displayText = Unsupported expression: a IS DISTINCT FROM b, type: IsDistinctFrom(Identifier(Ident { value: "a", quote_style: None }), Identifier(Ident { value: "b", quote_style: None })) (while in analyze select projection).
+ERROR 1105 (HY000) at line 39: Code: 1005, displayText = Unsupported expression: a IS NOT DISTINCT FROM b, type: IsNotDistinctFrom(Identifier(Ident { value: "a", quote_style: None }), Identifier(Ident { value: "b", quote_style: None })) (while in analyze select projection).
+ERROR 1105 (HY000) at line 45: Code: 1005, displayText = Unsupported expression: a IS DISTINCT FROM b, type: IsDistinctFrom(Identifier(Ident { value: "a", quote_style: None }), Identifier(Ident { value: "b", quote_style: None })) (while in analyze select projection).
+ERROR 1105 (HY000) at line 51: Code: 1005, displayText = Unsupported expression: a IS NOT DISTINCT FROM b, type: IsNotDistinctFrom(Identifier(Ident { value: "a", quote_style: None }), Identifier(Ident { value: "b", quote_style: None })) (while in analyze select projection).
 ==is_distinct_from_non_nullable_const_values==
-1
-0
-1
-0
 ==is_not_distinct_from_non_nullable_const_values==
-0
-1
-0
-1
 ==is_distinct_from_nullable_const_values==
-1
-1
-1
-1
-0
 ==is_not_distinct_from_nullable_const_values==
-0
-0
-0
-0
-1
 ==is_distinct_non_nullable_columns==
-0	0	0
-0	1	1
-1	0	1
-1	1	0
 ==is_not_distinct_non_nullable_columns==
-0	0	1
-0	1	0
-1	0	0
-1	1	1
 ==is_distinct_nullable_columns==
-0	NULL	1
-NULL	0	1
-NULL	NULL	0
 ==is_distinct_nullable_columns==
-0	NULL	0
-NULL	0	0
-NULL	NULL	1
```

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
@mergify update

**Comment by sundy-li** [comment]
I'll continue this pr to make this ready to merge.

**Pull request merged**